tests/test_multi_async.py

```diff
@@ -256,31 +260,39 @@ def test_014_003_restart_node1():
     node1.run()
 
     # node1 used to be primary, now demoted, and meanwhile node2 was primary
-    # so we can't decide to promote node1 now, it's stuck demoted
-    assert node1.wait_until_state(target_state="demoted")
-    assert node3.wait_until_state(target_state="report_lsn")
-
-    # ensure that node1 stays demoted
-    time.sleep(5)
-    assert node1.wait_until_state(target_state="demoted")
+    # node1 is assigned report_lsn and then is selected (only node with
+    # candidate priority > 0) ; and thus needs to go through fast_forward
+    assert node1.wait_until_state(target_state="fast_forward")
+    assert node1.wait_until_state(target_state="stop_replication")
+    assert node1.wait_until_state(target_state="wait_primary")
+    assert node3.wait_until_state(target_state="secondary")
 
 
 def test_014_004_restart_node2():
     node2.run()
 
-    assert node2.wait_until_state(target_state="wait_primary")
-
+    assert node2.wait_until_state(target_state="secondary")
     assert node3.wait_until_state(target_state="secondary")
-    assert node1.wait_until_state(target_state="secondary")
-
-    assert node2.wait_until_state(target_state="primary")
+    assert node1.wait_until_state(target_state="primary")
 
 
 #
 # Test 15 is like test 14, though we inverse the restarting of the failed
 # nodes, first the new primary and then very old primary.
 #
 #   node1        node2        node3
+#   primary      secondary    secondary
+#   <down>
+#   demoted      wait_primary secondary
+#                <down>
+#   demoted      demoted      report_lsn
+#                <up>
+#   demoted      wait_primary secondary
+#   <up>
+#   secondary    primary      secondary
+#
+#   ~~~
+#
```

Comment:
I think this new order should be instead of the one below

Reply:
Thanks, that was a leftover that should have been cleaned-up. Done now.